### PR TITLE
command/{cp, mv, rm}: add dry-run option

### LINKED ISSUE #90: add --dry-run option
Hi, 
it would be nice if you could add `--dry-run` to the command line options.

### PR DESCRIPTION
**Pull request by fbarotov**
Resolves #90

### CODE REVIEW

**Comment by sonmezonur** [comment]
Can we additionally add `--dry-run` option to run command? 
```
s5cmd run --dry-run commandfile
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+#### Features
+- Added `--dry-run` option for copy, move, and remove operations. It displays which commands will be executed without actually having a side effect. ([#90](https://github.com/peak/s5cmd/issues/90))
```
Comment (sonmezonur) [inline CHANGELOG.md:4]: Added `--dry-run` option for run, copy, move and remove operations.

File: command/cp.go [hunk 2]
```diff
@@ -70,6 +70,9 @@ Examples:
 
 	12. Perform KMS-SSE of the object(s) at the destination using customer managed Customer Master Key (CMK) key id
 		> s5cmd {{.HelpName}} -sse aws:kms -sse-kms-key-id <your-kms-key-id> s3://bucket/object s3://target-bucket/prefix/object
+
+	13. Check what s5cmd will do, without actually doing so (check mode option)
+		> s5cmd {{.HelpName}} -dry-run dir/ s3://bucket/
```
Comment (igungor) [inline command/cp.go:75]: `--dry-run`

Comment (igungor) [reply command/cp.go:75]: better drop `(check mode option)`

File: command/run.go [hunk 3]
```diff
@@ -32,10 +32,19 @@ Examples:
 		 > cat commands.txt | s5cmd {{.HelpName}}
 `
 
+var runCommandFlags = []cli.Flag{
+	&cli.BoolFlag{
+		Name:    "dry-run-all",
```
Comment (igungor) [inline command/run.go:37]: why not `--dry-run`?

Comment (fbarotov) [reply command/run.go:37]: normally, parent context flag values can be accessed from the child context. but if same flags are used they cannot, child context flag overrides to empty even if it is not set.

so in this case if  `--dry-run` flag is used for both `run` command and `mv`/`cp`. setting this flag in `run` has no effect for `mv`/`cp`

File: command/cp.go [hunk 4]
```diff
@@ -70,6 +70,9 @@ Examples:
 
 	12. Perform KMS-SSE of the object(s) at the destination using customer managed Customer Master Key (CMK) key id
 		> s5cmd {{.HelpName}} -sse aws:kms -sse-kms-key-id <your-kms-key-id> s3://bucket/object s3://target-bucket/prefix/object
+
+	13. Check what s5cmd will do, without actually doing so
+		> s5cmd {{.HelpName}} --dry-run dir/ s3://bucket/
```
Comment (ilkinulas) [inline command/cp.go:75]: In `dry-run` mode `cp` command prepares local destination which introduces side effects.

For example below command creates directory `~/TEMP/ilkin` if it does not exists.
```shell
s5cmd cp --dry-run  's3://test-bucket/files*'  ~/TEMP/ilkin/
```

Comment (fbarotov) [reply command/cp.go:75]: thanks for pointing it out

Comment (sonmezonur) [reply command/cp.go:75]: Can you add tests for upload and download operations? As far as I see, all dry-run tests are covering s3 copy (s3->s3). We can add assertions to show whether directories/files created.

File: command/cp.go [hunk 5]
```diff
@@ -585,15 +604,20 @@ func prepareLocalDestination(
...
+		} else if err := os.MkdirAll(dsturl.Dir(), os.ModePerm); err != nil {
 			return nil, err
 		}
 	}
 
 	if err == storage.ErrGivenObjectNotFound {
-		if err := os.MkdirAll(dsturl.Dir(), os.ModePerm); err != nil {
+		if dryRun {
+			// dryRun => no side effects.
+		} else if err := os.MkdirAll(dsturl.Dir(), os.ModePerm); err != nil {
```
Comment (igungor) [inline command/cp.go:617]: why not `if !dryRun`?

File: command/rm.go [hunk 6]
```diff
@@ -34,13 +34,22 @@ Examples:
 
 	4. Delete all matching objects and a specific object
 		 > s5cmd {{.HelpName}} s3://bucketname/prefix/* s3://bucketname/object1.gz
+
+	5. Check what s5cmd will do, without actually doing so (check mode option)
+		> s5cmd {{.HelpName}} -dry-run s3://bucket/prefix/*
```
Comment (igungor) [inline command/rm.go:39]: `--dry-run`

File: command/run.go (hunk 3 above)
Comment (igungor) [reply command/run.go:37]: hmm ok.

File: command/cp.go (hunk 4 above)
Comment (fbarotov) [reply command/cp.go:75]: @igungor yeah, it's addressed. thanks for the tip, will do as such from now on.

File: command/cp.go (hunk 5 above)
Comment (fbarotov) [reply command/cp.go:617]: @igungor yes, https://github.com/peak/s5cmd/pull/200#discussion_r469279834 is addressed

Comment (fbarotov) [reply command/cp.go:617]: > why not `if !dryRun`?

did, that to avoid having nested if statements

File: command/rm.go [hunk 7]
```diff
@@ -34,13 +34,22 @@ Examples:
 
 	4. Delete all matching objects and a specific object
 		 > s5cmd {{.HelpName}} s3://bucketname/prefix/* s3://bucketname/object1.gz
+
+	5. Check what s5cmd will do, without actually doing so (check mode option)
```
Comment (igungor) [inline command/rm.go:38]: drop (check mode option)

File: e2e/cp_test.go [hunk 8]
```diff
@@ -2701,3 +2702,77 @@ func TestCopyWithNoFollowSymlink(t *testing.T) {
...
+		),
+	}
+
+	workdir := fs.NewDir(t, t.Name(), folderLayout...)
+	defer workdir.Remove()
+
+	srcpath := filepath.ToSlash(workdir.Path())
+	dstpath := fmt.Sprintf("s3://%v/", bucket)
+
+	cmd := s5cmd("cp", "--dry-run", workdir.Path()+"/", dstpath)
```
Comment (igungor) [inline e2e/cp_test.go:2731]: also assert the filesystem that there's no change

File: e2e/cp_test.go [hunk 9]
```diff
@@ -2701,3 +2702,77 @@ func TestCopyWithNoFollowSymlink(t *testing.T) {
...
+	result := icmd.RunCmd(cmd)
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: equals("cp %v/c/file2.txt dir/c/file2.txt", srcpath),
+		1: equals("cp %v/file1.txt dir/file1.txt", srcpath),
+	}, sortInput(true))
+
+	// not even outermost directory should be created
```
Comment (igungor) [inline e2e/cp_test.go:2775]: assert s3 too

File: e2e/run_test.go [hunk 10]
```diff
@@ -232,3 +233,52 @@ func TestRunSpecialCharactersInPrefix(t *testing.T) {
...
+	filecontent := []string{
+		fmt.Sprintf("cp s3://%v/%s s3://%v/cp_%s", bucket, files[0], bucket, files[0]),
+		fmt.Sprintf("mv s3://%v/%s s3://%v/mv_%s", bucket, files[1], bucket, files[1]),
+		fmt.Sprintf("rm s3://%v/%s", bucket, files[2]),
+	}
+
+	file := fs.NewFile(t, "prefix", fs.WithContent(strings.Join(filecontent, "\n")))
+	defer file.Remove()
+
+	cmd := s5cmd("run", "--dry-run", file.Path())
```
Comment (igungor) [inline e2e/run_test.go:261]: could you add variations of this test?

- try with `s5cmd run --dry-run-all`
- try with `s5cmd run --dry-run-all` and the payload contains `cp --dry-run foo bar`
- try with `s5cmd run --dry-run` and the payload contains `cp --dry-run foo bar`

Comment (igungor) [reply e2e/run_test.go]: awesome 💯

**Comment by igungor** [comment]
#210 is the prefered implementation. Closing this one.

**Pull request closed without merging**